pytorch_toolkit/person_reidentification/engine/engine.py

```diff
@@ -0,0 +1,355 @@
+
+            if (epoch + 1) >= start_eval and eval_freq > 0 and (epoch + 1) % eval_freq == 0:
+                rank1 = self.test(
+                    epoch,
+                    testloader,
+                    dist_metric=dist_metric,
+                    normalize_feature=normalize_feature,
+                    visrank=visrank,
+                    visrank_topk=visrank_topk,
+                    save_dir=save_dir,
+                    use_metric_cuhk03=use_metric_cuhk03,
+                    ranks=ranks
+                )
+                self._save_checkpoint(epoch, rank1, save_dir)
+
+        elapsed = round(time.time() - time_start)
+        elapsed = str(datetime.timedelta(seconds=elapsed))
+        print('Elapsed {}'.format(elapsed))
+
+        if self.writer is not None:
+            self.writer.flush()
+            self.writer.close()
+
+    def train(self, epoch, max_epoch, trainloader, fixbase_epoch=0, open_layers=None, print_freq=10):
+        losses = AverageMeter()
+        reg_ow_loss = AverageMeter()
+        metric_losses = AverageMeter()
+        accs = AverageMeter()
+        batch_time = AverageMeter()
+        data_time = AverageMeter()
+
+        self.model.train()
+        if (epoch + 1) <= fixbase_epoch and open_layers is not None:
+            print('* Only train {} (epoch: {}/{})'.format(open_layers, epoch+1, fixbase_epoch))
+            open_specified_layers(self.model, open_layers)
+        else:
+            open_all_layers(self.model)
+
+        num_batches = len(trainloader)
+        start_time = time.time()
+        for batch_idx, data in enumerate(trainloader):
+            data_time.update(time.time() - start_time)
+
+            imgs, pids = self._parse_data_for_train(data)
+            imgs, pids = self._apply_batch_transform(imgs, pids)
+            if self.use_gpu:
+                imgs = imgs.cuda()
+                pids = pids.cuda()
+
+            self.optimizer.zero_grad()
+            if self.metric_losses is not None:
+                embeddings, outputs = self.model(imgs, get_embeddings=True)
+            else:
+                outputs = self.model(imgs)
+
+            loss = self._compute_loss(self.criterion, outputs, pids)
+
+            if (epoch + 1) > fixbase_epoch:
+                reg_loss = self.regularizer(self.model)
+                reg_ow_loss.update(reg_loss.item(), pids.size(0))
+                loss += reg_loss
+
+            if self.metric_losses is not None:
+                self.metric_losses.writer = self.writer
+                self.metric_losses.init_iteration()
+                metric_loss = self.metric_losses(embeddings, pids, epoch, epoch * num_batches + batch_idx)
+                self.metric_losses.end_iteration()
+                loss += metric_loss
+
+            loss.backward()
+            self.optimizer.step()
+
+            losses.update(loss.item(), pids.size(0))
+            metric_losses.update(metric_loss.item(), pids.size(0))
```

Comment:
@sovrasov ,
Here is a runtime error if `self.metric_losses` is not used and related to not declared variable `metric_loss`.